include/wil/registry_basic.h

```diff
@@ -0,0 +1,1540 @@
+    const reg_view_details::reg_view_nothrow regview{key};
+    return regview.open_key(path, hkey, access);
+}
+#endif
+
+inline HRESULT create_key_nothrow(HKEY key, _In_ PCWSTR path, ::wil::unique_hkey& hkey, ::wil::registry::key_access access = ::wil::registry::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr)
+{
+    const reg_view_details::reg_view_nothrow regview{key};
+    return regview.create_key(path, hkey, access, security_descriptor);
+}
+
+#if defined(__WIL_WINREG_STL)
+inline HRESULT create_key_nothrow(HKEY key, _In_ PCWSTR path, ::wil::shared_hkey& hkey, ::wil::registry::key_access access = ::wil::registry::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr)
+{
+    const reg_view_details::reg_view_nothrow regview{key};
+    return regview.create_key(path, hkey, access, security_descriptor);
+}
+#endif
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+inline size_t get_child_key_count(HKEY key)
+{
+    DWORD numSubKeys{};
+    THROW_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+        key, nullptr, nullptr, nullptr, &numSubKeys, nullptr,
+        nullptr, nullptr, nullptr, nullptr, nullptr, nullptr));
+    return numSubKeys;
+}
+
+inline size_t reg_get_child_value_count(HKEY key)
+{
+    DWORD numSubValues{};
+    THROW_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+        key, nullptr, nullptr, nullptr, nullptr, nullptr,
+        nullptr, &numSubValues, nullptr, nullptr, nullptr, nullptr));
+    return numSubValues;
+}
+#endif
+
+inline HRESULT get_child_key_count_nothrow(HKEY key, _Out_ DWORD* numSubKeys) WI_NOEXCEPT
+{
+    RETURN_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+        key, nullptr, nullptr, nullptr, numSubKeys, nullptr,
+        nullptr, nullptr, nullptr, nullptr, nullptr, nullptr));
+    return S_OK;
+}
+
+inline HRESULT reg_get_child_value_count_nothrow(HKEY key, _Out_ DWORD* numSubValues) WI_NOEXCEPT
+{
+    RETURN_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+        key, nullptr, nullptr, nullptr, nullptr, nullptr,
+        nullptr, numSubValues, nullptr, nullptr, nullptr, nullptr));
+    return S_OK;
+}
+
+// Registry Set* functions, both throwing and non-throwing
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+inline void set_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD data)
+{
+    const reg_view_details::reg_view regview{key};
+    return regview.set_value(subkey, value_name, data);
+}
+
+inline void set_value_dword(HKEY key, _In_opt_ PCWSTR value_name, DWORD data)
+{
+    return set_value_dword(key, nullptr, value_name, data);
+}
+
+inline void set_value_qword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, uint64_t data)
+{
+    const reg_view_details::reg_view regview{key};
+    return regview.set_value(subkey, value_name, data);
+}
+
+inline void set_value_qword(HKEY key, _In_opt_ PCWSTR value_name, uint64_t data)
+{
+    return set_value_qword(key, nullptr, value_name, data);
+}
+
+inline void set_value_string(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _In_ PCWSTR data)

```

Comment:
We need to support variants that take `std::wstring` (and potentially other string types/string_view).

Why? Registry strings can have embedded nulls 🤣, so a PCWSTR can lead to data loss.

Reply:
sadly, std::wstring cannot have embedded nulls either (at least to be conformant). IMHO, shoving something with embedded nulls as a REG_SZ should just go through the raw byte* version - it should stick out badly as "I'm doing something super-strange and not recommended, so I have to use the raw API to do it, not a friendly API" (I don't want to suggest this is a good idea by having friendly support for it).

Reply:
I think that's a good point. We shouldn't make it easy to do Bad Things.

However, I know that C++/WinRT is deliberate to recommend string_views in some cases when there's interop with HSTRINGs, in part because HSTRINGs can have embedded nulls, right?

I recommend at least including a note indicating our behavior for this edge case (truncation, right?). Optimally we'd have some sort of variant call pattern that returns a vector of bytes or something (or an HSTRING or something?)